compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -45,25 +47,53 @@ impl Function {
 }
 
 impl Loops {
-    fn hoist_loop_invariants(self, function: &mut Function) {
+    fn hoist_loop_invariants(mut self, function: &mut Function) {
         let mut context = LoopInvariantContext::new(function);
 
-        for loop_ in self.yet_to_unroll.iter() {
+        // The loops should be sorted by the number of blocks.
+        // We want to access outer nested loops first, which we do by popping
+        // from the top of the list.
+        while let Some(loop_) = self.yet_to_unroll.pop() {
             let Ok(pre_header) = loop_.get_pre_header(context.inserter.function, &self.cfg) else {
                 // If the loop does not have a preheader we skip hoisting loop invariants for this loop
                 continue;
             };
-            context.hoist_loop_invariants(loop_, pre_header);
+
+            context.hoist_loop_invariants(&loop_, pre_header);
+
+            context.defined_in_loop.clear();
```

Comment:
Wouldn't it be better done in `set_values_defined_in_loop` instead, since _set_ typically means to discard what is there and replace it with something else, and `hoist_loop_invariants` will start with that? Then there would be no chance to forget calling it. 

Why don't we have to similarly clear `loop_invariants`? IIUC we go from outer loops to inner loops, so for example if we have:
```
let a = todo!()
'outer: loop {
   let b = todo!();
   'inner: loop {
      let c = todo!();
   }
}
```
then the `yet_to_unroll` vec has `['inner, 'outer]` and we start with `'outer`, which means we'll have `b` and `c` in `defined_in_loop`, and `a` in `loop_invariants`. If we didn't clear before processing `'inner`, it then `'inner` would think it defined `b`. OTOH if `a` was an invariant for `'outer` then it is also an invariant for `'inner`. Is that it?

Then, `'inner` inserts `b` as invariant; if we now had a sibling for `outer` then it would see both `a` and `b` as invariants, but presumably it would never access `b` and never be confused by it.

Wouldn't it be safer to clear it though?

Reply:
> Wouldn't it be better done in set_values_defined_in_loop instead, since set typically means to discard what is there and replace it with something else, and hoist_loop_invariants will start with that? Then there would be no chance to forget calling it.

I moved this to the top of `set_values_defined_in_loop`, I agree it is cleaner.

Reply:
> OTOH if `a` is an invariant for `'outer` then it is also an invariant for `'inner`, so we might as well keep it. Is that it?

This is why I did not clear loop invariants. It is not _necessary_ to clear them like the variables defined in a loop. Although, it should not affect the logic of the pass to clear them per loop like we do `defined_in_loop`. I decided to not clear them as to avoid clearing and reinserting loop invariants. I guess this shouldn't be a big cost though and we may as well be clearer about what is "per loop". I'll clear them as well.